Firestore/core/src/util/executor_libdispatch.mm

```diff
@@ -243,91 +113,184 @@ void MarkDone() {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    async_tasks_.insert(task);
+  }
+
+  dispatch_async_f(dispatch_queue_, task, InvokeAsync);
 }
+
 void ExecutorLibdispatch::ExecuteBlocking(Operation&& operation) {
-  DispatchSync(dispatch_queue(), std::move(operation));
+  HARD_ASSERT(
+      GetCurrentQueueLabel() != GetQueueLabel(dispatch_queue_),
+      "Calling DispatchSync on the current queue will lead to a deadlock.");
+
+  auto task = Task::Create(this, std::move(operation));
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    async_tasks_.insert(task);
+  }
+
+  dispatch_sync_f(dispatch_queue_, task, InvokeSync);
 }
 
-DelayedOperation ExecutorLibdispatch::Schedule(const Milliseconds delay,
-                                               TaggedOperation&& operation) {
+DelayedOperation ExecutorLibdispatch::Schedule(Milliseconds delay,
+                                               Tag tag,
+                                               Operation&& operation) {
   namespace chr = std::chrono;
   const dispatch_time_t delay_ns = dispatch_time(
       DISPATCH_TIME_NOW, chr::duration_cast<chr::nanoseconds>(delay).count());
 
   // Ownership is fully transferred to libdispatch -- because it's impossible
-  // to truly cancel work after it's been dispatched, libdispatch is
-  // guaranteed to outlive the executor, and it's possible for work to be
-  // invoked by libdispatch after the executor is destroyed. Executor only
-  // stores an observer pointer to the operation.
-  TimeSlot* time_slot = nullptr;
-  TimeSlotId time_slot_id = 0;
-  RunSynchronized(this, [this, delay, &operation, &time_slot, &time_slot_id] {
-    time_slot_id = NextId();
-    time_slot = new TimeSlot{this, delay, std::move(operation), time_slot_id};
-    schedule_[time_slot_id] = time_slot;
-  });
-
-  dispatch_after_f(delay_ns, dispatch_queue(), time_slot,
-                   TimeSlot::InvokedByLibdispatch);
-
-  return DelayedOperation{[this, time_slot_id] {
+  // to truly cancel work after it's been dispatched, libdispatch is guaranteed
+  // to outlive the executor, and it's possible for work to be invoked by
+  // libdispatch after the executor is destroyed. The Executor only stores an
+  // observer pointer to the operation.
+  Task* task = nullptr;
+  TimePoint target_time = MakeTargetTime(delay);
+  Id id = 0;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    id = NextIdLocked();
+    task = Task::Create(this, target_time, tag, id, std::move(operation));
+    async_tasks_.insert(task);
+    schedule_[id] = task;
+  }
+
+  dispatch_after_f(delay_ns, dispatch_queue_, task, InvokeAsync);
+
+  return DelayedOperation(this, id);
+}
+
+void ExecutorLibdispatch::Complete(Task* task) {
+  bool should_release = false;
+  {
+    TASK_TRACE("Executor::Complete %s task %s", this, task);
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    auto found = async_tasks_.find(task);
+    if (found != async_tasks_.end()) {
```

Comment:
Can you add a comment on why a `Task` might not be found?

Reply:
I can't think of a good reason why it might not be found--concurrent access to the executor is forbidden and tasks should only complete themselves once.

Given that everything else in this vicinity handles duplicate invocation it seemed sensible to do so here as well. If you prefer, I can make the assertion that the task is found. I'm slightly wary of this though, since the consequences are dire and the behavior on not found seems reasonable.